tofhir-server/src/main/scala/io/tofhir/server/service/TerminologySystemService.scala

```diff
@@ -0,0 +1,134 @@
+package io.tofhir.server.service
+
+import com.typesafe.scalalogging.LazyLogging
+import io.tofhir.engine.model.{FhirMappingJob, LocalFhirTerminologyServiceSettings}
+import io.tofhir.server.model.TerminologySystem
+import io.tofhir.server.service.job.IJobRepository
+import io.tofhir.server.service.project.IProjectRepository
+import io.tofhir.server.service.terminology.ITerminologySystemRepository
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class TerminologySystemService(terminologySystemRepository: ITerminologySystemRepository, mappingJobRepository: IJobRepository, projectRepository: IProjectRepository) extends LazyLogging {
+
+  /**
+   * Get all TerminologySystem metadata from the TerminologySystem repository
+   * @return List of TerminologySystem metadata
+   */
+  def getAllMetadata: Future[Seq[TerminologySystem]] = {
+    terminologySystemRepository.getTerminologySystemsMetadata
+  }
+
+  /**
+   * Create a new TerminologySystem
+   * @param terminologySystem The TerminologySystem will be created
+   * @return Created TerminologySystem
+   */
+  def createTerminologySystem(terminologySystem: TerminologySystem): Future[TerminologySystem] = {
+    terminologySystemRepository.createTerminologySystem(terminologySystem)
+  }
+
+  /**
+   * Get the TerminologySystem by id
+   *
+   * @param id id of the TerminologySystem
+   * @return The TerminologySystem with given id
+   */
+  def getTerminologySystem(id: String): Future[Option[TerminologySystem]] = {
+    terminologySystemRepository.getTerminologySystem(id)
+  }
+
+  /**
+   * Update the TerminologySystem
+   * @param id TerminologySystem id
+   * @param terminologySystem TerminologySystem to update
+   * @return Updated TerminologySystem
+   */
+  def updateTerminologySystem(id: String, terminologySystem: TerminologySystem): Future[TerminologySystem] = {
+
+    // Get project ids from project repository
+    val projectIds: Future[Seq[String]] = projectRepository.getAllProjects.map(_.map(_.id))
+
+    projectIds.flatMap { projectIds =>
+        Future.traverse(projectIds) { projectId =>
+
+        // Get jobs of given project
+        val jobs: Future[Seq[FhirMappingJob]] = mappingJobRepository.getAllJobs(projectId)
+
+        jobs.map { jobs =>
+          jobs
+            .filter { job =>
+              job.terminologyServiceSettings.isDefined && job.terminologyServiceSettings.get.isInstanceOf[LocalFhirTerminologyServiceSettings]
```

Comment:
This filter can also include the if check we used in line 70. Further, that part is duplicated in line 111,  therefore you can implement a simple method to be passed into the filter method here.